Fix DB action timers housekeeping.

hjoliver opened the pull request:
<!-- Complete this Pull Request template. -->

<!-- Significant PRs should address an existing Issue. Choose one: -->
These changes close #3590 

Bug allows `task_action_timers` DB table housekeeping to be skipped, so the table grows forever, and causes screeds of warning messages on restart.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
<!-- choose one: -->
- [x] Appropriate tests are included (unit and/or functional).
<!-- choose one: -->
- [ ] Appropriate change log entry included.
<!-- choose one: -->
- [x] No documentation update required. (Not user-facing)

hjoliver commented:
(Might as well go to the alpha-2 milestone at this point).

hjoliver commented on cylc/flow/suite_db_mgr.py:
```diff
@@ -355,19 +355,18 @@ def put_suite_template_vars(self, template_vars):
 
     def put_task_event_timers(self, task_events_mgr):
         """Put statements to update the task_action_timers table."""
-        if task_events_mgr.event_timers:
```
This variable, `event_timers` in the task event manager, does not hold task poll and try timers (they are task proxy attributes) which are also written to the `task_action_timers` DB table.  It only holds job log retrieval timers (so suites that need remote job log retrieval presumably do get correct DB housekept correctly - this might explain why no one has noticed this problem in operational suites?).

oliver-sanders replied:
Due to this line:

https://github.com/cylc/cylc-flow/blob/35b160453601fbb7cd92a83e6f3251b90b758b9b/cylc/flow/suite_db_mgr.py#L204

It would appear that there is no consequence to this line: 

```python
self.db_deletes_map[self.TABLE_TASK_ACTION_TIMERS].append({})
```

So we should be good.

hjoliver replied:
> It would appear that there is no consequence to this line

Huh, by experiment, this change results in the table been emptied, and then repopulated with only the current timers.

oliver-sanders replied:
Aah so an empty "delete map" means "refresh".

hjoliver commented:
Simplest example of the problem:
```
[scheduling]
   cycling mode = integer
   initial cycle point = 1
   max active cycle points = 1
   [[graph]]
        P1 = "foo[-P1] => foo"
```
Run this and watch the table grow indefinitely from ICP onward:
```bash
$ watch -n 1 "sqlite3 /home/oliverh/cylc-run/foo/log/db 'select * from task_action_timers'"

1|foo|"poll_timer"|["tuple", [[1, "submitted"]]]|[]|0||
1|foo|["try_timers", "retrying"]|null|[]|0||
1|foo|["try_timers", "submit-retrying"]|null|[]|0||
2|foo|"poll_timer"|["tuple", [[1, "submitted"]]]|[]|0||
2|foo|["try_timers", "retrying"]|null|[]|0||
2|foo|["try_timers", "submit-retrying"]|null|[]|0||
3|foo|"poll_timer"|["tuple", [[1, "submitted"]]]|[]|0||
3|foo|["try_timers", "retrying"]|null|[]|0||
3|foo|["try_timers", "submit-retrying"]|null|[]|0||
4|foo|"poll_timer"|["tuple", [[1, "submitted"]]]|[]|0||
4|foo|["try_timers", "retrying"]|null|[]|0||
4|foo|["try_timers", "submit-retrying"]|null|[]|0||
...
```

hjoliver commented:
Damn it, functional tests are the bane of my life.  Two tests that check for shutdown on deletion of the run directory are failing on this branch because now the housekeeping evidently gets done before the run-dir health check.

hjoliver commented:
Not sure what to do about that - the health check is now an asynchronous plug-in, so we can't guarantee order.

Seems to me those shutdown tests are badly designed. the scheduler shuts down if the run directory is found to have disappeared, but it will also shut down before that check is done if any operation to read or write from the missing run directory fails.

oliver-sanders commented:
Looks good to me, not sure how to get around the test failures.

We're going to have to do something about the functional tests at some point, they are making progress very difficult.

hjoliver commented:
I've modified `grep_ok` to allow arbitrary `grep` options, and fixed the functional tests by allowing either valid failure mode on deleted run dir: `grep -E "(FAIL1|FAIL2)" (extended regex search).

hjoliver commented:
@kinow if you approve this we can merge it once tests pass (I'll take @oliver-sanders comment as approval if tests fixed: "Looks good to me, not sure how to get around the test failures.")

kinow commented:
Sure @hjoliver . I think pycodestyle found something in the latest changes though. Wil review once GitHub actions tests pass.

hjoliver commented:
Yeah just pushed a shell style fix.

hjoliver commented:
> Yeah just pushed a shell style fix.

Gack, that didn't work too well :grimacing:

kinow commented on tests/shutdown/14-no-dir-check.t:
```diff
@@ -36,9 +36,13 @@ ln -s "$(basename "${SUITE_NAME}")" "${SYM_SUITE_RUND}"
+# Possible failure modes:
+# - health check detects missing run directory
+# - DB housekeeping cannot access DB because run directory missing
+# - (TODO: if other failure modes show up, add to the list here!)
```
:+1:
